generator/internal/data_model.h

```diff
@@ -58,13 +62,17 @@ struct DataModel {
         internal::ProtoNameToCppName(method->output_type()->full_name());
   }
 
-  static void PrintMethods(pb::ServiceDescriptor const* service,
-                           std::map<std::string, std::string> vars, Printer& p,
-                           char const* tmplt) {
+  static void PrintMethods(
+      pb::ServiceDescriptor const* service,
+      std::map<std::string, std::string> vars, Printer& p, char const* tmplt,
+      std::function<bool(pb::MethodDescriptor const*)> predicate =
```

Comment:
This is fine for this PR, but we should start to think about how we will handle printing methods once the requirements become more complex, i.e. we will need different templates for simple unary, streaming, and also paginated and longrunning methods. It may be that we just abandon the PrintMethods function entirely, I'm not sure.

Reply:
I will put my neck on the line and predict that most of the heavy lifting can be done via the templates passed in and control flow done by the caller. That seems straightforward and procedural, but I don't have a good mental image for the more complex operations. It also makes the templates potentially more complex then the surrounding code, which sounds more unpleasant to maintain.

The other obvious design is to lift manipulating the services into the caller, which would copy/partition the pointers to the MethodDescriptors into e.g. lroMethods, simpleMethods, streamingMethods lists. That more tightly couples the template and partitioning, and we could aggregate the sync and async variations of a method to save on calls to SetMethodVars. It still seems like manipulating the template is going to be most of the work, though, so this may be a bit of a wash.

Reply:
It almost feels like we want to go full blown OO and make a GeneratedFile class with Header and CC child classes and instantiate one per file that we want to create. We always want to create a list of includes, always want a list of namespaces, for headers we always want an include guard, we want to generate a filepath based on the file type and service descriptor name.
The main generator loop would then iterate over the services and create file objects that are responsible for partitioning methods into unary/streaming/osv.

Reply:
I wouldn't be opposed to this! But, I would council caution at this stage. In my initial PR, I had some abstractions around templating, and they turned out to be wrong and we all decided to remove them. I think we are now getting more clarity about what is the right level/type of abstraction, but I'd suggest we hold off a bit longer until the generator stabilizes somewhat, and then reassess. WDYT?

Reply:
I agree with the urge for additional caution. That having been said, there's already an annoying amount of duplication and manual manipulation that a properly structured class hierarchy might improve. I'll finish this PR with the current structure and then explore some designs and maybe put together a proposal doc.